### PR TITLE
Fixed #25406 -- Removed exception hiding in PostgreSQL test database …

### PR DESCRIPTION
**Pull request by felixxm**
…creation / cloning during `--keepdb`. Ticket [25406](https://code.djangoproject.com/ticket/25406).

I'm going to prepare separate PR with similar modification in the MySQL backend.

### CODE REVIEW

**Comment by timgraham** [comment]
Do you want to try adding tests using mocking like the similar Oracle PR?

**Comment by felixxm** [comment]
Yes, I'll try to prepare tests.

**Comment by felixxm** [comment]
I added tests.

**Comment by felixxm** [comment]
@MarkusH Thanks for the review.

Currently we mock database creation, hence if we add
```diff
diff --git a/tests/backends/test_creation.py b/tests/backends/test_creation.py
index 759d192..48f216f 100644
--- a/tests/backends/test_creation.py
+++ b/tests/backends/test_creation.py
@@ -115,6 +115,8 @@ class PostgreSQLDatabaseCreationTests(SimpleTestCase):
                     # SystemExit is raised if the user answers "no" to the
                     # prompt asking if it's okay to delete the test database.
                     creation._create_test_db(verbosity=0, autoclobber=False, keepdb=False)
+            with mock.patch('builtins.input', return_value='yes'):
+                 creation._create_test_db(verbosity=0, keepdb=False)
             # "Database already exists" error is ignored when keepdb is on
             creation._create_test_db(verbosity=0, autoclobber=False, keepdb=True)
         # Simulate test database creation raising unexpected error
```
than `DROP DATABASE` will always fail with `django.db.utils.ProgrammingError: database "test_test_djangotest" does not exist`. I'm not sure if it's worth adding such test.

File: django/db/backends/postgresql/creation.py [hunk 1]
```diff
@@ -28,29 +28,42 @@ def sql_table_creation_suffix(self):
...
+        try:
+            super()._execute_test_db_creation(cursor, parameters, keepdb)
+        except Exception as e:
+            exc_msg = 'database %s already exists' % parameters['dbname']
+            if exc_msg not in str(e):
+                # All errors except "database already exists" cancel tests
+                sys.stderr.write("Got an error creating the test database: %s\n" % e)
+                sys.exit(2)
+            elif not keepdb:
+                # if we want to keep the db, then no need to raise exception
```
Comment (timgraham) [inline django/db/backends/postgresql/creation.py]: I'd change this to: `If the database should be kept, ignore "database already exists".`

File: django/db/backends/base/creation.py [hunk 2]
```diff
@@ -148,27 +148,28 @@ def _get_test_db_name(self):
...
         """
-        suffix = self.sql_table_creation_suffix()
-
         test_database_name = self._get_test_db_name()
-
-        qn = self.connection.ops.quote_name
+        test_db_param = {
+            'dbname': self.connection.ops.quote_name(test_database_name),
+            'suffix': self.sql_table_creation_suffix(),
+        }
```
Comment (timgraham) [inline django/db/backends/base/creation.py]: I'd chop this blank line since the } on its own line is providing whitespace.

File: django/db/backends/base/creation.py [hunk 3]
```diff
@@ -148,27 +148,28 @@ def _get_test_db_name(self):
...
             try:
-                cursor.execute(
-                    "CREATE DATABASE %s %s" % (qn(test_database_name), suffix))
+                self._execute_test_db_creation(cursor, test_db_param, keepdb)
             except Exception as e:
                 # if we want to keep the db, then no need to do any of the below,
                 # just return and skip it all.
                 if keepdb:
                     return test_database_name
-
```
Comment (timgraham) [inline django/db/backends/base/creation.py]: To keep the diff a bit cleaner, I wouldn't make this unrelated whitespace change.

File: django/db/backends/base/creation.py [hunk 4]
```diff
@@ -148,27 +148,28 @@ def _get_test_db_name(self):
             return self.connection.settings_dict['TEST']['NAME']
         return TEST_DATABASE_PREFIX + self.connection.settings_dict['NAME']
 
+    def _execute_test_db_creation(self, cursor, parameters, keepdb=False):
```
Comment (timgraham) [inline django/db/backends/base/creation.py]: maybe: `_execute_create_test_db`?

File: django/db/backends/base/creation.py [hunk 5]
```diff
@@ -148,27 +148,28 @@ def _get_test_db_name(self):
...
     def _create_test_db(self, verbosity, autoclobber, keepdb=False):
         """
         Internal implementation - creates the test db tables.
         """
-        suffix = self.sql_table_creation_suffix()
-
         test_database_name = self._get_test_db_name()
-
-        qn = self.connection.ops.quote_name
+        test_db_param = {
```
Comment (timgraham) [inline django/db/backends/base/creation.py]: param -> params

File: tests/backends/test_creation.py [hunk 6]
```diff
@@ -54,6 +54,15 @@ def test_custom_test_name_with_test_prefix(self):
 @unittest.skipUnless(connection.vendor == 'postgresql', "PostgreSQL-specific tests")
 class PostgreSQLDatabaseCreationTests(SimpleTestCase):
 
+    def _execute_raise_database_already_exists(self, cursor, parameters, keepdb=False):
+        raise DatabaseError("database %s already exists" % parameters['dbname'])
```
Comment (timgraham) [inline tests/backends/test_creation.py]: single quotes?

File: tests/backends/test_creation.py [hunk 7]
```diff
@@ -54,6 +54,15 @@ def test_custom_test_name_with_test_prefix(self):
 @unittest.skipUnless(connection.vendor == 'postgresql', "PostgreSQL-specific tests")
 class PostgreSQLDatabaseCreationTests(SimpleTestCase):
 
+    def _execute_raise_database_already_exists(self, cursor, parameters, keepdb=False):
```
Comment (timgraham) [inline tests/backends/test_creation.py]: Would it be better to put the helper methods near the tests they're used in? (I'm not sure whether or not they'll be reused for other tests or if there are any guidelines about this ordering).

Comment (felixxm) [inline tests/backends/test_creation.py]: I don't think they are reusable. I moved them above the test.

**Comment by felixxm** [comment]
@timgraham Thanks for review. I updated PR.

File: tests/backends/test_creation.py [hunk 8]
```diff
@@ -95,6 +95,35 @@ def test_sql_table_creation_suffix_with_encoding_and_template(self):
...
         self.check_sql_table_creation_suffix(settings, '''WITH ENCODING 'UTF8' TEMPLATE "template0"''')
 
+    def _execute_raise_database_already_exists(self, cursor, parameters, keepdb=False):
+        raise DatabaseError('database %s already exists' % parameters['dbname'])
+
+    def _execute_raise_permission_denied(self, cursor, parameters, keepdb=False):
+        raise DatabaseError('permission denied to create database')
+
+    def patch_test_db_creation(self, execute_test_db_creation):
+        return mock.patch.object(BaseDatabaseCreation, '_execute_test_db_creation', execute_test_db_creation)
```
Comment (timgraham) [inline tests/backends/test_creation.py]: _execute_create_test_db to fix the tests.
